### ISSUE TITLE
no positional info used for SemanticTransformer

### ISSUE THREAD

**Issue by drwangxian** [lucidrains/audiolm-pytorch#286]
A big problem: no positional information was used for SemanticTransformer

Below code was used to define the SemanticTransformer:

```
semantic_transformer = SemanticTransformer(
    num_semantic_tokens = wav2vec.codebook_size,
    dim = 1024,
    depth = 6,
    flash_attn = True
)
```

The first few lines of the SemanticTransformer is as follows:

```
class SemanticTransformer(nn.Module):
    @beartype
    def __init__(
        self,
        *,
        dim,
        depth,
        num_semantic_tokens,
        heads = 8,
        attn_dropout = 0.,
        ff_dropout = 0.,
        t5_name = DEFAULT_T5_NAME,
        cond_dim = None,
        has_condition = False,
        audio_text_condition = False,
        cond_as_self_attn_prefix = False,
        cond_drop_prob = 0.5,
        grad_shrink_alpha = 0.1,
        rel_pos_bias = True,
        flash_attn = False,
        **kwargs
    ):
        super().__init__()
        rel_pos_bias = rel_pos_bias and not flash_attn
```

Consequently, when flash_attn = True,  rel_pos_bias was always False. Therefore, no positional information was used.

**Comment by lucidrains** [comment]
ah yes, that's because learned relative positional bias is not compatible with flash attention

do you plan on using this repo for your research? what type of audio are you training on? if it is speech, there are [better alternatives](https://github.com/lucidrains/e2-tts-pytorch) these days. if not, i can spend some time modernizing this repo, just let me know

**Comment by drwangxian** [comment]
Need not modernize the code for now. I will switch to E2 TTS. I am not for research. I just want to learn some practical skills in speech. I found for E2 TTS, a checkpoint is available, so it can help a lot.

**Comment by lucidrains** [comment]
@drwangxian use f5-tts, which is based off e2-tts and has a multilingual open sourced model available

good luck with your research

**Comment by drwangxian** [comment]
Thanks for bringing f5-tts to my attention!

Your practice of using no positional embedding seemlingly comes from "Transformer Language Models without Positional Encodings Still Learn Positional Information" (https://arxiv.org/abs/2203.16634). This paper states that causal masking can somehow help the model to be position aware.

**Comment by lucidrains** [comment]
@drwangxian yes it did, but that paper ended up not being picked up in practice

from what i've heard, it still performs way worse compared to having positional embedding. still an interesting paper tho

i'll update it to rotary embeddings soon
